src/python/pants/core/goals/update_build_files.py

```diff
@@ -436,4 +494,5 @@ def rules():
         # after all our deprecation fixers.
         UnionRule(RewrittenBuildFileRequest, FormatWithBlackRequest),
         UnionRule(RewrittenBuildFileRequest, FormatWithYapfRequest),
+        # UnionRule(RewrittenBuildFileRequest, FormatWithRuffRequest),
```

Comment:
Note - this line should be uncommented, but doing so causes a lot of rule graph errors for reasons unknown. Snippet of the full output:

```
ValueError: Encountered 42 rule graph errors:
  No installed rules return the type FormatWithBlackRequest, and it was not provided by potential callers of @rule(pants.core.goals.update_build_files:369:format_build_file_with_black(FormatWithBlackRequest, Black) -> RewrittenBuildFile, gets=[Get(Snapshot, [CreateDigest]), Get(Lockfile, [Resolve]), Get(LoadedLockfile, [LoadedLockfileRequest]), Get(VenvPex, [PexRequest]), Get(ConfigFiles, [ConfigFilesRequest]), Get(Digest, [MergeDigests]), Get(ProcessResult, [VenvPexProcess]), Get(Snapshot, [Digest]), Get(DigestContents, [Digest])]).
    If that type should be computed by a rule, ensure that that rule is installed.
    If it should be provided by a caller, ensure that it is included in any relevant Query or Get.
  No installed rules return the type FormatWithYapfRequest, and it was not provided by potential callers of @rule(pants.core.goals.update_build_files:335:format_build_file_with_yapf(FormatWithYapfRequest, Yapf) -> RewrittenBuildFile, gets=[Get(Snapshot, [CreateDigest]), Get(Lockfile, [Resolve]), Get(LoadedLockfile, [LoadedLockfileRequest]), Get(VenvPex, [PexRequest]), Get(ConfigFiles, [ConfigFilesRequest]), Get(Digest, [MergeDigests]), Get(ProcessResult, [VenvPexProcess]), Get(Snapshot, [Digest]), Get(DigestContents, [Digest])]).
    If that type should be computed by a rule, ensure that that rule is installed.
    If it should be provided by a caller, ensure that it is included in any relevant Query or Get.
  No installed rules return the type PexCliProcess, and it was not provided by potential callers of @rule(pants.backend.python.util_rules.pex_cli:114:setup_pex_cli_process(PexCliProcess, PexPEX, PexEnvironment, PythonBuildStandaloneBinary, PythonNativeCodeSubsystem.EnvironmentAware, GlobalOptions, PexSubsystem, PythonSetup) -> Process, gets=[Get(Digest, [CreateDigest]), Get(Digest, [MergeDigests])]).
    If that type should be computed by a rule, ensure that that rule is installed.
    If it should be provided by a caller, ensure that it is included in any relevant Query or Get.
  No installed rules return the type PexProcess, and it was not provided by potential callers of @rule(pants.backend.python.util_rules.pex:1149:setup_pex_process(PexProcess, PexEnvironment) -> Process, gets=[Get(Digest, [MergeDigests])]).
    If that type should be computed by a rule, ensure that that rule is installed.
    If it should be provided by a caller, ensure that it is included in any relevant Query or Get.
  No installed rules return the type RenameDeprecatedFieldsRequest, and it was not provided by potential callers of @rule(pants.core.goals.update_build_files:465:maybe_rename_deprecated_fields(RenameDeprecatedFieldsRequest) -> RewrittenBuildFile, gets=[Get(FixedBUILDFile, [RenameFieldsInFileRequest])]).
    If that type should be computed by a rule, ensure that that rule is installed.
    If it should be provided by a caller, ensure that it is included in any relevant Query or Get.
  No installed rules return the type RenameDeprecatedTargetsRequest, and it was not provided by potential callers of @rule(pants.core.goals.update_build_files:437:maybe_rename_deprecated_targets(RenameDeprecatedTargetsRequest) -> RewrittenBuildFile, gets=[Get(FixedBUILDFile, [RenameTargetsInFileRequest])]).
    If that type should be computed by a rule, ensure that that rule is installed.
    If it should be provided by a caller, ensure that it is included in any relevant Query or Get.
  No installed rules return the type VenvPexProcess, and it was not provided by potential callers of @rule(pants.backend.python.util_rules.pex:1239:setup_venv_pex_process(VenvPexProcess, PexEnvironment) -> Process, gets=[Get(Digest, [MergeDigests])]).
    If that type should be computed by a rule, ensure that that rule is installed.
    If it should be provided by a caller, ensure that it is included in any relevant Query or Get.
    ```

Reply:
I posted some thoughts in [slack](https://pantsbuild.slack.com/archives/C0D7TNJHL/p1706774174006329?thread_ts=1705342639.696869&cid=C0D7TNJHL). I think the short description of the problem is that somewhere there's a missing rule (somewhere in the chain involving FormatWithRuffRequest) that causes a projection through the union to fail (since not all members implement the projection), which causes all the other graphs to not close. eg `Get(T, Parent)` can't use a rule Child0->T because Child1->T doesn't exist, so Pants finds dangling rule Child0->T and says that nothing provides Child0.
To help isolate the fault, you could try turning off backends, that'll at least reduce the number of rules you're seeing.